[Li Ting] iP

Goh-Li-Ting opened the pull request:
# ChatBuddy

> "Time management is the key to success." - Unknown

ChatBuddy helps you to take note of your tasks. It's
- ~~easy~~ VERY EASY to use
- text-based and suitable for people who prefer typing over clicking buttons
- supports different categories of tasks

## Instructions
To start your to do list, you only need to

1. Download it from [here](https://github.com/Goh-Li-Ting/ip/releases/tag/A-Jar).
1. Start the application.
1. Add your tasks.

And it is **completely free of charge**!
_Donations and publicity are appreciated!_ 😄 

## Features
- [x] Add tasks to do list
- [x] Categorising tasks by _to do_, _deadline_ and _event_
- [ ] Filter tasks by dates (coming soon)

Here's the `main` method from `ChatBuddy`:
```Java
public static void main(String[] args) {
    new ChatBuddy("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Goh-Li-Ting committed:
Add greet and exit

Goh-Li-Ting committed:
Add ability to echo commands

Goh-Li-Ting committed:
Add ability to add tasks to list and display list

Goh-Li-Ting committed:
Add ability to mark as done and not done

Goh-Li-Ting committed:
Add todo, event and deadline tasks

Goh-Li-Ting committed:
Add text UI testing

Goh-Li-Ting committed:
Add error handling

Goh-Li-Ting committed:
Add ability to delete tasks

Goh-Li-Ting committed:
Add ability to save tasks and retrieve saved tasks

Goh-Li-Ting committed:
Merge branch 'branch-Level-7'

Goh-Li-Ting committed:
Add datetime format to deadline and event

Goh-Li-Ting committed:
Add ability to get deadlines/events within a week

Goh-Li-Ting committed:
Merge branch 'branch-Level-8'

Goh-Li-Ting committed:
Abstract out classes for OOP

Goh-Li-Ting committed:
Merge branch 'branch-A-MoreOOP'

Goh-Li-Ting committed:
Organise classes into packages

Goh-Li-Ting committed:
Merge branch 'branch-A-Packages'

Goh-Li-Ting committed:
Merge branch 'add-gradle-support'

Goh-Li-Ting committed:
Set up gradle to run ChatBuddy using gradle

Goh-Li-Ting committed:
Merge branch 'branch-A-Gradle'

Goh-Li-Ting committed:
Add JUnit tests

Goh-Li-Ting committed:
Merge branch 'branch-A-JUnit'

Goh-Li-Ting committed:
Package app as JAR file

Goh-Li-Ting committed:
Add JavaDoc comments

Goh-Li-Ting committed:
Add ability to find tasks by keyword

Goh-Li-Ting committed:
Change code to comply with coding standard

Goh-Li-Ting committed:
Merge branch 'branch-A-JavaDoc'

Goh-Li-Ting committed:
Merge branch 'branch-A-CodingStandard'

Goh-Li-Ting committed:
Merge branch 'branch-Level-9'

Goh-Li-Ting committed:
Add checkstyle

Goh-Li-Ting committed:
Merge branch 'branch-A-CheckStyle'

Sebtey commented on src/main/java/chatbuddy/command/Command.java:
```diff
@@ -0,0 +1,29 @@
+     *
+     * @return False by default.
+     */
+    public boolean isExit() {
```
Great usage of inheritance here! 👍

Sebtey commented on src/main/java/chatbuddy/command/Command.java:
```diff
@@ -0,0 +1,29 @@
+import chatbuddy.ui.Ui;
+
+/** Command is an abstract class that represents commands that can be executed. */
+public abstract class Command {
```
Hey, I noticed that all of the Command child classes have a common field "COMMAND_WORD". 

Just a suggestion, would it be a better alternative to have that field encapsulated under Command abstract class?

Sebtey commented on src/main/java/chatbuddy/ui/Ui.java:
```diff
@@ -0,0 +1,110 @@
+ * The Ui class handles interactions with the user.
+ * It reads inputs from the users and output messages to the user.
+ */
+public class Ui {
```
Perhaps you can consider creating an attribute to represent the buffer in front of each lines printed onto the console?

I believe that might help minimize the chance for a line printed by Ui to have a different buffer size at the front.

Sebtey reviewed (commented):
Overall, LGTM. I really like how succinct your code is and how you fully utilise OOP concepts in this IP! 👍 I find the documentations and testings really concise as well.

Suggested a few areas that I believe can be further encapsulated in my review, hope you can kindly consider them as well!

A1WAYSD commented on src/main/java/chatbuddy/TaskList.java:
```diff
@@ -0,0 +1,155 @@
+     */
+    public Task markTaskAsNotDone(int taskNum) throws ChatBuddyException {
+        // check for valid task number
+        checkValidTaskNumber(taskNum);
```
I like how you generalize the checkTask into a method and the additional helpful comment written:D

A1WAYSD commented on src/main/java/chatbuddy/task/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+ */
+public class Deadline extends Task {
+
+    /** The formatter used for date inputs. */
+    private static final DateTimeFormatter FORMATTER_DATE_INPUT = DateTimeFormatter.ofPattern("dd/MM/yyyy");
+    /** The formatter used for date outputs. */
+    private static final DateTimeFormatter FORMATTER_DATE_OUTPUT = DateTimeFormatter.ofPattern("dd MMM yyyy");
```
Perhaps indentation for wrapped lines should be 8 spaces?

A1WAYSD commented on src/main/java/chatbuddy/task/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+     * @return True if the deadline is due within a week, false otherwise.
+     */
+    @Override
+    public boolean isWithinAWeek() {
```
Perhaps should use a verb and can change the method into checkWithinAWeek?

A1WAYSD commented on src/main/java/chatbuddy/task/Event.java:
```diff
@@ -0,0 +1,70 @@
+public class Event extends Task {
+
+    /** The formatter used for datetime inputs. */
+    private static final DateTimeFormatter FORMATTER_DATETIME_INPUT = DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm");
+    /** The formatter used for datetime outputs. */
+    private static final DateTimeFormatter FORMATTER_DATETIME_OUTPUT = DateTimeFormatter.ofPattern("dd MMM yyyy HHmm");
```
I've noticed the same issue in other places too..

A1WAYSD reviewed (commented):
LGTM!! The test cases are very thorough, include both trivial and non-trivial cases.
Like how you split into different actions and closely follow OOP principles! Good job.

Goh-Li-Ting committed:
Add GUI

Goh-Li-Ting committed:
Merge branch 'branch-Level-10'

Goh-Li-Ting committed:
Add assertions

No assertions are used in TaskList.

Assertions can help to identity bugs in code.

Add assertions to TaskList to check that
* the task number is valid if the code reaches the point to mark as done or not done
* the new task is successfully added to the back of the taskList
* the deleted task is no longer inside the taskList at the end of the function

Such assertions help to check that the functions work as intended.

Goh-Li-Ting committed:
Merge pull request #1 from Goh-Li-Ting/branch-A-Assertions

Add assertions to TaskList.java

Goh-Li-Ting committed:
Merge branch 'master' into branch-A-CodeQuality

Goh-Li-Ting committed:
Improve code quality

The current code is not of good quality.
Complicated expressions are used and SLAP is not followed for some methods.

Improve code quality in Parser.java, Storage.java and Ui.java.
* Extract code to follow SLAP and make code more readable
* Change some complicated expressions to named booleans

Improving code quality makes the code easier to read.

Goh-Li-Ting committed:
Merge pull request #2 from Goh-Li-Ting/branch-A-CodeQuality

Improve code quality

Goh-Li-Ting committed:
Add ability to add tags to tasks

Goh-Li-Ting committed:
Merge branch 'branch-C-Tagging'

Goh-Li-Ting committed:
Update user guide

Goh-Li-Ting committed:
Fix user guide formatting issue

Goh-Li-Ting committed:
Add product screenshot

Goh-Li-Ting committed:
Change save data to save on every change

Data is written to the file only after the ExitCommand is ran or when
the user closes the interface.

This makes it risky as data can be lost if the ExitCommand is not ran
or the application did not close properly.

Saving the data to the file after every command that causes that data
to change will make it safer as data is saved even if the application
runs into unexpected issues.

Saving data after necessary commands ensure that the data is saved.

Goh-Li-Ting committed:
Change find command to be case-insensitive

Goh-Li-Ting committed:
Improve test cases

Goh-Li-Ting committed:
Change ExitCommand to close the application
